File: haskell/src/Day6.hs
module Day6 (runPart1, runPart2) where

import qualified Data.Set as Set
import Util

input :: String
input = "../inputs/day6.txt"

readInput :: IO [[Set.Set Char]]
readInput = do
  fileContents <- readFile input
  fileContents
    |> lines
    |> partitionBy (== "")
    |> filter (/= [""])
    |> map (map Set.fromList)
    |> pure

runPart1 :: IO ()
runPart1 = do
  groups <- readInput
  groups
    |> map (foldr1 Set.union)
    |> map length
    |> sum
    |> print

runPart2 :: IO ()
runPart2 = do
  groups <- readInput
  groups
    |> map (foldr1 Set.intersection)
    |> map length
    |> sum
    |> print

File: rust/src/aoc/day6.rs
use itertools::Itertools;
use std::collections::HashSet;
use std::fs;
use std::io;
use std::io::prelude::*;

const INPUT: &str = "../inputs/day6.txt";

fn read_input() -> Vec<Vec<HashSet<char>>> {
    let input = fs::File::open(INPUT).unwrap();
    let file_contents = io::BufReader::new(input);
    let groups = file_contents
        .lines()
        .map(|l| l.unwrap())
        .group_by(|l| l == "");

    groups
        .into_iter()
        .filter(|(is_empty, _)| !*is_empty)
        .map(|(_, group)| group.map(|g| g.chars().collect()).collect())
        .collect()
}

#[allow(dead_code)]
pub fn run_part_1() {
    let answers = read_input()
        .into_iter()
        .map(|group| {
            group
                .into_iter()
                .fold_first(|s1, s2| &s1 | &s2)
                .unwrap()
                .len()
        })
        .sum::<usize>();

    println!("{:?}", answers)
}

#[allow(dead_code)]
pub fn run_part_2() {
    let answers = read_input()
        .into_iter()
        .map(|group| {
            group
                .into_iter()
                .fold_first(|s1, s2| &s1 & &s2)
                .unwrap()
                .len()
        })
        .sum::<usize>();

    println!("{:?}", answers)
}
